src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs

```diff
@@ -67,113 +69,153 @@ public static ParseUDFsResult ParseUDFsScript(string script, CultureInfo loc = n
+        private Token TokPlainEat(TokKind tid)
+        {
+            if (_curs.TidCur == tid)
+            {
+                return _curs.TokMove();
+            }
+
+            return null;
+        }
+
         private ParseUDFsResult ParseUDFs(string script)
         {
             var udfs = new List<UDF>();
 
+            // <root> ::= (<udf> ';')*
             while (_curs.TokCur.Kind != TokKind.Eof)
             {
-                //var args = new HashSet<UDFArg>();
-                var args = new Dictionary<string, UDFArg>();
                 ParseTrivia();
-
-                // Store the name of the UDF.
-                var ident = TokEat(TokKind.Ident);
-                if (ident == null)
+                if (!ParseUDF(udfs))
                 {
-                    CreateError(_curs.TokCur, TexlStrings.ErrUDF_MissingIdentifier);
-                    break;
+                    return new ParseUDFsResult(udfs, _errors);
                 }
 
-                if (TokEat(TokKind.ParenOpen) == null)
+                ParseTrivia();
+                
+                if (TokEat(TokKind.Semicolon) == null)
                 {
-                    CreateError(_curs.TokCur, TexlStrings.ErrUDF_MissingOpenParen);
                     break;
                 }
+            }
 
-                ParseTrivia();
-
-                // Parse arguments.
-                while (_curs.TokCur.Kind != TokKind.ParenClose)
-                {
-                    ParseTrivia();
-                    var varIdent = TokEat(TokKind.Ident);
-                    ParseTrivia();
-                    if (TokEat(TokKind.Colon) == null)
-                    {
-                        CreateError(_curs.TokCur, TexlStrings.ErrUDF_MissingColon);
-                        break;
-                    }
-
-                    ParseTrivia();
-                    var varType = TokEat(TokKind.Ident);
-                    ParseTrivia();
-                    if (args.ContainsKey(varIdent.ToString()))
-                    {
-                        CreateError(_curs.TokCur, TexlStrings.ErrUDF_DuplicateArgName);
-                        return new ParseUDFsResult(udfs, _errors);
-                    }
-
-                    args.Add(varIdent.ToString(), new UDFArg(varIdent.As<IdentToken>(), varType.As<IdentToken>()));
-                    if (_curs.TokCur.Kind != TokKind.ParenClose && _curs.TokCur.Kind != TokKind.Comma)
-                    {
-                        CreateError(_curs.TokCur, TexlStrings.ErrUDF_MissingComma);
-                        return new ParseUDFsResult(udfs, _errors);
-                    }
-                    else if (_curs.TokCur.Kind == TokKind.Comma)
-                    {
-                        TokEat(TokKind.Comma);
-                    }
-                }
+            return new ParseUDFsResult(udfs, _errors);
+        }
 
-                TokEat(TokKind.ParenClose);
+        private bool ParseUDFArgs(HashSet<UDFArg> args)
```

Comment:
Make this a ref variable. I know it's not required for mutable collections, but it helps the user know that the collection is being modified.

In fact, the new hash set can be created inside this function, and the param marked as "out" instead of "ref".

Reply:
Changed